src/hooks/search/filter.ts

```diff
@@ -32,17 +32,50 @@ const useFilter = (
     }, [filterItems, router?.isReady, router.query]);
 
     const hasFiltersSet = !!filters.length;
+
+    const subFilterTypeToMainFilterType: Record<string, string> = {};
+    filterItems?.forEach((filterItem: { key: string; value: FilterItem[] }) => {
+        filterItem.value.forEach((filter: FilterItem) => {
+            if (!filter.subFilters?.length) return;
+
+            filter.subFilters.forEach((subfilter: FilterItem) => {
+                const subfilterType = subfilter.type as string;
+                subFilterTypeToMainFilterType[subfilterType] =
+                    filter.type as string;
+            });
+        });
+    });
+
+    // refactor the map out of itemInFilters()
+    // so it is created once instead of searchData.length times
+    const allFiltersTypeMap: { [type: string]: FilterItem[] } = {};
+    filters.forEach((filter: FilterItem) => {
+        if (!filter.type) return;
+
+        let filterType = filter.type;
+        if (filterType in subFilterTypeToMainFilterType) {
+            filterType = subFilterTypeToMainFilterType[filterType];
```

Comment:
This is the root of the bug. Subfilters have different types. Below is an example. 

```
{
  "name": "Event",
  "type": "ContentType",
  "subFilters": [
    {
      "name": "User Group Meetup",
      "type": "EventType",
      "subFilters": [],
      "count": 30
    },
    {
      "name": "Industry Event",
      "type": "EventType",
      "subFilters": [],
      "count": 12
    }
  ],
  "count": 42
}
```

`allFiltersTypeMap` previously only used `type` as a key, so User Group Meetup was created with a type `EventType` in `allFiltersTypeMap`. 

When applying the [core filter logic](https://github.com/mongodb/devcenter/blob/89f5bc34aa4305bf777978fa9f4a8f2807afa8ec/src/hooks/search/utils.ts#L300), `User Group Meetup` and `Events` are considered different groups and filtered using AND logic.

The above example can be generalized to all other pairs of sub-filters and main-filters.

Reply:
Filter logic here perhaps is also best to be encapsulated within the backend instead of being processed by the frontend, for both performance and code organization purposes.

Reply:
at that time we dint have backend to support complex search apis in backend

Reply:
Agreed